### PR TITLE
SubgridScaleTurbulence-WIP

### PR DESCRIPTION
**Pull request by asraero**
Subgrid Scale (SGS) Turbulence models

1) Introduces new module for subgrid-scale turbulence models
- [x] Standard Smagorinsky-Lilly Model (SSM)

2) Drivers 
- General changes: Uses MoistThermodynamics for saturation adjustment and wave-speed correction in moist atmospheres
- Viscous stresses 
- No-flux boundary conditions 
- Additional diagnostic outputs
(Docs will be included when this is switched to a full PR)

List 
- [x] Density current (Dry) (test for SGS filters)  https://doi.org/10.1002/fld.165017010
- [x] Rising bubble (Dry, Moist) https://doi.org/10.1175/1520-0469(1993)050<1865:BCEWAS>2.0.CO;2
- [ ] 2D Squall Line (e.g. Marras et al. JCP 2013) https://doi.org/10.1016/j.jcp.2013.06.006

Future PRs will add to this - 
- Anisotropic minimum dissipation
- Parameter-free (residual-based) Dyn-SGS method
Other relevant branch: https://github.com/asraero/CLIMA/tree/dycoms

### CODE REVIEW

File: src/Atmos/Dycore/src/VanillaAtmosDiscretizations_cuda.jl [hunk 1]
```diff
@@ -0,0 +1,1242 @@
+# FIXME: Add link to https://github.com/paranumal/libparanumal here and in
```
Comment (simonbyrne) [inline src/Atmos/Dycore/src/VanillaAtmosDiscretizations_cuda.jl]: I'm assuming you didn't mean to add this file?

**Comment by glwagner** [comment]
Sorry, I didn't mean to edit your text --- I didn't realize that simply clicking on the box would check it off!

File: src/Atmos/Dycore/src/VanillaAtmosDiscretizations_cuda.jl (hunk 1 above)
Comment (asraero) [reply src/Atmos/Dycore/src/VanillaAtmosDiscretizations_cuda.jl]: Yes - this is unintentional

**Comment by asraero** [comment]
@lcw @smarras79 : We can generalize the spatial equivalent length scales to be computed in terms of the metric terms (generalises the module to spherical coordinate system).

File: test/DGmethods/compressible_Navier_Stokes/dc_smagorinsky_sgs.jl [hunk 2]
```diff
@@ -0,0 +1,586 @@
...
+@inline function wavespeed(n, Q, aux, t, P, u, v, w, ρinv, q_liq, T, θ)
+    gravity::eltype(Q) = grav
+    @inbounds begin 
+        ρ, U, V, W, E, QT = Q[_ρ], Q[_U], Q[_V], Q[_W], Q[_E], Q[_QT]
+        x,y,z = aux[_a_x], aux[_a_y], aux[_a_z]
+        u, v, w = ρinv * U, ρinv * V, ρinv * W
+        e_int = (E - (U^2 + V^2+ W^2)/(2*ρ) - ρ * gravity * y) / ρ
+        q_tot = QT / ρ
+        TS = PhaseEquil(e_int, q_tot, ρ)
+        (n[1] * u + n[2] * v + n[3] * w) + soundspeed_air(TS)
```
Comment (jkozdon) [inline test/DGmethods/compressible_Navier_Stokes/dc_smagorinsky_sgs.jl]:
```suggestion
abs(n[1] * u + n[2] * v + n[3] * w) + soundspeed_air(TS)
```

File: test/DGmethods/compressible_Navier_Stokes/rtb_smagorinsky_sgs.jl [hunk 3]
```diff
@@ -0,0 +1,584 @@
...
+@inline function wavespeed(n, Q, aux, t, P, u, v, w, ρinv, q_liq, T, θ)
+    gravity::eltype(Q) = grav
+    @inbounds begin 
+        ρ, U, V, W, E, QT = Q[_ρ], Q[_U], Q[_V], Q[_W], Q[_E], Q[_QT]
+        x,y,z = aux[_a_x], aux[_a_y], aux[_a_z]
+        u, v, w = ρinv * U, ρinv * V, ρinv * W
+        e_int = (E - (U^2 + V^2+ W^2)/(2*ρ) - ρ * gravity * y) / ρ
+        q_tot = QT / ρ
+        TS = PhaseEquil(e_int, q_tot, ρ)
+        (n[1] * u + n[2] * v + n[3] * w) + soundspeed_air(TS)
```
Comment (jkozdon) [inline test/DGmethods/compressible_Navier_Stokes/rtb_smagorinsky_sgs.jl]:
```suggestion
abs(n[1] * u + n[2] * v + n[3] * w) + soundspeed_air(TS)
```

Comment (asraero) [reply test/DGmethods/compressible_Navier_Stokes/rtb_smagorinsky_sgs.jl]: fixed

File: test/DGmethods/compressible_Navier_Stokes/dycoms.jl [hunk 4]
```diff
@@ -1,611 +0,0 @@
...
-    datap          = spl_pinit(xvert)
-    dataq          = dataq * 1.0e-3
-    
-    randnum   = rand(1)[1] / 100
-
-    θ_liq = datat
-    q_tot = dataq + randnum * dataq
-    P     = datap    
-    T     = air_temperature_from_liquid_ice_pottemp(θ_liq, P, PhasePartition(q_tot))
-    ρ     = air_density(T, P)
```
Comment (jkozdon) [inline test/DGmethods/compressible_Navier_Stokes/dycoms.jl]: Not sure if this file was supposed to be deleted, but in the dycoms driver should this be
```
    ρ     = air_density(T, P, PhasePartition(q_tot))
```

Comment (smarras79) [reply test/DGmethods/compressible_Navier_Stokes/dycoms.jl]: This file is now obsolete and should not be merged to the main branch.

Comment (jkozdon) [reply test/DGmethods/compressible_Navier_Stokes/dycoms.jl]: Same question about density applies to all the dycoms drivers I have seen

File: src/Atmos/Parameterizations/SubgridScaleTurbulence/SubgridScaleTurbulence.jl [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+    Sij are components of the strain-rate tensor, and SijSij is the tensor-inner-product
+
+  Compute components of strain-rate tensor from velocity gradient terms provided in 
+  driver. Note that the gradient terms are computed in Cartesian space even on the spherical
+  grid, with velocities and other key variables projected onto spherical coordinates for 
+  visualisation 
+
+  Dij = ∇u .................................................. [1]
+  Sij = 1/2 (∇u + (∇u)ᵀ) .....................................[2]
+  τij = 2 * ν_e * Sij ........................................[3]
```
Comment (tapios) [inline src/Atmos/Parameterizations/SubgridScaleTurbulence/SubgridScaleTurbulence.jl]: Can we use the sign convention here that gives us momentum fluxes, i.e., `\tau_ij = -2 \nu_e S_ij` (as in design document)?

File: src/Atmos/Parameterizations/SubgridScaleTurbulence/SubgridScaleTurbulence.jl [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+  """
+  function strainrate_tensor_components(dudx, dudy, dudz, dvdx, dvdy, dvdz, dwdx, dwdy, dwdz)
+    # Assemble components of the strain-rate tensor 
+    S11, = dudx
+    S12  = (dudy + dvdx) / 2
+    S13  = (dudz + dwdx) / 2
+    S22  = dvdy
+    S23  = (dvdz + dwdy) / 2
+    S33  = dwdz
+    SijSij = S11^2 + S22^2 + S33^2 + 2 * (S12^2 + S13^2 + S23^2)
```
Comment (tapios) [inline src/Atmos/Parameterizations/SubgridScaleTurbulence/SubgridScaleTurbulence.jl]: Maybe `S2` instead of `SijSij` (or some other way of indicating a square) is easier?

File: src/Atmos/Parameterizations/SubgridScaleTurbulence/SubgridScaleTurbulence.jl [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+  doi = {10.1175/1520-0493(1963)091<0099:GCEWTP>2.3.CO;2},
+  URL = {https://doi.org/10.1175/1520-0493(1963)091<0099:GCEWTP>2.3.CO;2},
+  eprint = {https://doi.org/10.1175/1520-0493(1963)091<0099:GCEWTP>2.3.CO;2}
+  }
+  """
+  function standard_smagorinsky(SijSij, Δsqr)
+    # Eddy viscosity is a function of the magnitude of the strain-rate tensor
+    # This is for use on both spherical and cartesian grids. 
+    DFloat = eltype(SijSij)
+    ν_e::DFloat = sqrt(2.0 * SijSij) * C_smag * C_smag * Δsqr
```
Comment (tapios) [inline src/Atmos/Parameterizations/SubgridScaleTurbulence/SubgridScaleTurbulence.jl]: `C_smag^2` (and similarly for squares elsewhere?)

**Comment by asraero** [comment]
Reason for closure: Components moved to AtmosModel framework with #426 , #397. Anisotropic Minimum Dissipation is WIP in https://github.com/asraero/CLIMA/tree/sgs-amd-dgmodel.
